### PR TITLE
:running: Test Cluster conversions with object fuzzer

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
An initial implementation of an idea sparked from a conversation with @ncdc a while back. This PR adds a Test (only for Cluster for now) to test the conversion from v1alpha3 -> v1alpha2 -> v1alpha3.

If the code looks good, I'll add the rest in other PRs, and probably remove the tests for restore.

This change caught we were dropping `Status.FailureDomains` and `Spec.Paused`, so 🎉 

/assign @ncdc @detiber

### CODE REVIEW

File: api/v1alpha2/conversion_test.go [hunk 1]
```diff
@@ -17,15 +17,39 @@ limitations under the License.
...
 
+func TestFuzzyConversion(t *testing.T) {
+	g := NewWithT(t)
+
+	scheme := runtime.NewScheme()
+	g.Expect(AddToScheme(scheme)).To(Succeed())
+	g.Expect(v1alpha3.AddToScheme(scheme)).To(Succeed())
+	fuzzer := utilconversion.GetFuzzer(scheme)
+
+	for i := 0; i < rand.Intn(100); i++ {
```
Comment (ncdc) [inline api/v1alpha2/conversion_test.go]: You could probably convert this into a helper function that takes in `interface{}`s for src, dst, so in the api packages you'd just have to set up a scheme and call the helper func with the pairings you want to test (v1a2.Machine, v1a3.Machine; v1a2.Cluster, v1a3.Cluster; ...)

Comment (vincepri) [reply api/v1alpha2/conversion_test.go]: ptal

**Comment by ncdc** [comment]
@vincepri do you want to get this in, then follow up with more tests, or do them all together?

**Comment by vincepri** [comment]
I tried to add the other objects, and they're failing the test, because I have to finish some reviews, let's get this in now and I'll finish it up later this afternoon

File: util/conversion/conversion.go [hunk 2]
```diff
@@ -50,3 +62,28 @@ func UnmarshalData(from metav1.Object, to interface{}) (bool, error) {
 	delete(from.GetAnnotations(), DataAnnotation)
 	return true, nil
 }
+
+// GetFuzzer returns a new fuzzer to be used for testing.
+func GetFuzzer(scheme *runtime.Scheme) *fuzz.Fuzzer {
+	return fuzzer.FuzzerFor(
+		fuzzer.MergeFuzzerFuncs(metafuzzer.Funcs),
```
Comment (detiber) [inline util/conversion/conversion.go]: Is this actually fuzzing values in our types? iirc, we need to define our own fuzzing methods for custom types similar to this: https://github.com/kubernetes/kubernetes/blob/550fb1bfc3c773febbb0811262fd7a8f7847bb20/pkg/apis/core/fuzzer/fuzzer.go

File: util/conversion/conversion.go [hunk 3]
```diff
@@ -50,3 +62,28 @@ func UnmarshalData(from metav1.Object, to interface{}) (bool, error) {
 	delete(from.GetAnnotations(), DataAnnotation)
 	return true, nil
 }
+
+// GetFuzzer returns a new fuzzer to be used for testing.
```
Comment (ncdc) [inline util/conversion/conversion.go]: Shouldn't these belong in a _test.go file?

File: util/conversion/conversion.go (hunk 2 above)
Comment (ncdc) [reply util/conversion/conversion.go]: I'd have to track down what gofuzz is doing, but I don't think it's required but rather if you want to customize fuzzing?

**Comment by wfernandes** [comment]
When I run `go test -v ./... -race -count 2 -run TestFuzzy` I get it to fail consistently
```
--- FAIL: TestFuzzyConversion (0.01s)
    --- FAIL: TestFuzzyConversion/for_Cluster (0.00s)
        conversion.go:86:
            {"metadata":{"name":"òCʃǩŐ濾3Ò譭ż\"檗\u003cJ蓅ƢÃ兂H","generateName":"埵ƘȾĞÏǋ","namespace":"+«D/#á羿üŊɱ","selfLink":"ck'剚ƤÄǢúOI亷幡ɰ","uid":"!眞嘟Ƿ真À馾UĝĚ@Ğ扬ŧĺz櫟","resourceVersion":"15689593688535292584","generation":-877665905178434740,"creationTimestamp":null,"deletionGracePeriodSeconds":-4671736085913427863,"annotations":{"\\Ʊ滸a訽樣餤}Ǩ瘇ôɶ":"Ơñ"},"ownerReferences":[{"apiVersion":"up鳅欀ƀr,遭I","kind":"蜷gC侨","name":"枫遌¥Ɍ?!)?ǹ","uid":"¢ 悵ɂńƵGɱXXƜE餋"}],"finalizers":["ʥǷơ]N蝳遇莀垘輱"],"clusterName":"劋鏋@鎉tB篼Ʀǐ½eƈ巆ɐ錍v薕團"},"spec":{"paused":true,"clusterNetwork":{"apiServerPort":-451928821,"services":{"cidrBlocks":["ƭø"]},"serviceDomain":"Ŧi艤x'磺ŦCuUsá6ǲ翦ȧ;åĩ"},"controlPlaneEndpoint":{"host":"

            A: ","port":-1202100984},"infrastructureRef":{"kind":"ŨþM古ƢFF贔ox镋纋ĿÉ","namespace":"甃绛vdªXďP","name":"ª河ɵ","uid":"ƒɍřQB幯ǩƱ誄,åīȏǼxfƇƗ吋","apiVersion":"縇屽đ*贈V橻[$fƙƜ曲Ƌ7ȸɾg\"","resourceVersion":"窻Ų","fieldPath":"Ŕ(ǍƆȬ"}},"status":{"phase":"l\"纀绱Ǣ","infrastructureReady":false,"controlPlaneInitialized":true}}

            B: 中","port":2091358496},"infrastructureRef":{"kind":"ŨþM古ƢFF贔ox镋纋ĿÉ","namespace":"甃绛vdªXďP","name":"ª河ɵ","uid":"ƒɍřQB幯ǩƱ誄,åīȏǼxfƇƗ吋","apiVersion":"縇屽đ*贈V橻[$fƙƜ曲Ƌ7ȸɾg\"","resourceVersion":"窻Ų","fieldPath":"Ŕ(ǍƆȬ"}},"status":{"phase":"l\"纀绱Ǣ","infrastructureReady":false,"controlPlaneInitialized":true}}
```

I'm not sure if it is the last char of `host` or the ports.

**Comment by ncdc** [comment]
Ouch! Definitely need to look into this.

/hold

File: util/conversion/conversion.go (hunk 3 above)
Comment (vincepri) [reply util/conversion/conversion.go]: I don't think I can import them if they're in a test file

Comment (vincepri) [reply util/conversion/conversion.go]: I can try though, maybe I'm misremembering

**Comment by vincepri** [comment]
It seems that the port is not being converted as well properly, this is probably because we're going from int64 -> int32

File: util/conversion/conversion.go (hunk 3 above)
Comment (vincepri) [reply util/conversion/conversion.go]: yeah confirmed, it fails with

```
/Users/vince/go/src/sigs.k8s.io/cluster-api/api/v1alpha2/conversion_test.go:37:23: undefined: "sigs.k8s.io/cluster-api/util/conversion".FuzzTestFunc
FAIL	sigs.k8s.io/cluster-api/api/v1alpha2 [build failed]
```
if it's part of the test files

**Comment by ncdc** [comment]
I don't think it's that. It only fails when count>1. Feels like a pointer is being reused somewhere.

**Comment by vincepri** [comment]
If I increase the iterations it fails consistently here as well, I put it to 10k and it fails

**Comment by ncdc** [comment]
Oh wait, was just able to get it to fail with count=1 (I also changed the for loop to be a fixed 100 instead of a random value)

**Comment by ncdc** [comment]
Ok, what's happening in at least one case is we don't convert between control plane endpoint (a3) and apiendpoints (a2) if either the host or port is the zero value. But the fuzzer doesn't know about that, so if a3.controlPlaneEndpoint.host="" and port=$someNumber, it doesn't round trip.

**Comment by vincepri** [comment]
That + found an issue that we weren't properly using copies between each iteration

**Comment by vincepri** [comment]
Maybe the `IsZero` should return true only if both of the values are zero, seems a little strange that we return true if either one is empty?

**Comment by ncdc** [comment]
Sure, although it's not exactly valid to have one set and one empty. Should add webhook validation (separate PR).

**Comment by vincepri** [comment]
Fixed that bit in this PR

**Comment by ncdc** [comment]
/hold cancel
ran with count=50 (50*1000 loop iterations) and it all passed.

File: api/v1alpha3/cluster_types.go [hunk 4]
```diff
@@ -162,7 +162,7 @@ type APIEndpoint struct {
 
 // IsZero returns true if either the host or the port are zero values.
```
Comment (ncdc) [inline api/v1alpha3/cluster_types.go]: Update comment 😄

File: util/conversion/conversion.go (hunk 2 above)
Comment (ncdc) [reply util/conversion/conversion.go]: Circling back, yes, this will fuzz custom structs.

File: api/v1alpha3/cluster_types.go (hunk 4 above)
Comment (vincepri) [reply api/v1alpha3/cluster_types.go]: done

**Comment by ncdc** [comment]
LGTM. Y'all ok with this @detiber @wfernandes?

**Comment by vincepri** [comment]
Upped the iterations to 10k given that for larger objects might be needed, and it has very minimal impact on overall test run time

File: util/conversion/conversion.go (hunk 2 above)
Comment (vincepri) [reply util/conversion/conversion.go]: Yeah it'll work fine with custom structs. Although if you want to have some special fuzzing logic for some types, you can register more functions

**Comment by detiber** [comment]
lgtm

**Comment by ncdc** [comment]
lint timeout

**Comment by ncdc** [comment]
File an issue to reenable gomnd & fix magic numbers in the codebase?

**Comment by vincepri** [comment]
I'd rather disable it long term, no need to enable it imo?

**Comment by ncdc** [comment]
🤷‍♂

**Pull request merged**